### PR TITLE
Add support for creating reports from an event and posting

### PR DESCRIPTION
**Pull request by jepebe**
There are several unknowns with regard to how the general case of event and reports should be handeled. At the moment the solution is quite specific.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -55,6 +55,9 @@ async def main():
         reports = await toadr3.get_reports(session, VTN_URL, token)
         for report in reports:
             print(f"Report ID: {report.id} - {report.report_name}")
+            
+        report = toadr3.create_report(events[0])
```
Comment (jepebe) [inline README.md:59]: Probably missing some parameters

File: toadr3/report.py [hunk 2]
```diff
@@ -125,3 +127,113 @@ def payload_descriptors(self) -> list[ReportPayloadDescriptor] | None:
...
+    }
+
+    if report_name is not None:
+        data["reportName"] = report_name
+
+    if event.report_descriptors is None or len(event.report_descriptors) == 0:
+        raise ValueError("event does not have any report_descriptors.")
+
+    report_descriptor = None
+    for rd in event.report_descriptors:
```
Comment (millitza) [inline toadr3/report.py:196]: Can multiple report descriptors match?

File: toadr3/report.py [hunk 3]
```diff
@@ -125,3 +127,113 @@ def payload_descriptors(self) -> list[ReportPayloadDescriptor] | None:
...
+        if rd.payload_type == report_type:
+            report_descriptor = rd
+            break
+
+    if report_descriptor is None:
+        raise ValueError(f"event does not have a report_descriptor for {report_type}.")
+
+    data["payloadDescriptors"] = [{"payloadType": report_type}]
+
+    if "RESOURCE_NAME" not in event.targets:
```
Comment (millitza) [inline toadr3/report.py:206]: Maybe better to move this up to all the other validators as it does not require any values from the payload descriptors.

File: toadr3/reports.py [hunk 4]
```diff
@@ -5,6 +5,70 @@
...
+    if report is None:
+        raise ValueError("report is required")
+
+    headers = {}
+    if access_token is not None:
+        headers["Authorization"] = f"Bearer {access_token.token}"
+
+    vtn_url = vtn_url.rstrip("/")
+
+    async with session.post(f"{vtn_url}/reports", headers=headers, json=report) as response:
```
Comment (millitza) [inline toadr3/reports.py:46]: Don't you need to serialize the report?

File: toadr3/report.py [hunk 5]
```diff
@@ -125,3 +127,113 @@ def payload_descriptors(self) -> list[ReportPayloadDescriptor] | None:
     def resources(self) -> list[ReportData]:
         """The resources of the report."""
         return self._resources
+
+
+def create_report(
```
Comment (millitza) [inline toadr3/report.py:132]: If I understand correctly, we send a reports to the VTN to update them about the power usage. Could you add some documentation about the use of the report and the fields? I think report_values contains the measured power values. But other than that I have some issues understanding what this is used for and how. This is 100% my lack of knowledge about ADR, but that makes it difficult to use the library :)

File: toadr3/report.py [hunk 6]
```diff
@@ -125,3 +127,113 @@ def payload_descriptors(self) -> list[ReportPayloadDescriptor] | None:
...
+        "objectType": "REPORT",
+        "programID": event.program_id,
+        "eventID": event.id,
+        "clientName": client_name,
+    }
+
+    if report_name is not None:
+        data["reportName"] = report_name
+
+    if event.report_descriptors is None or len(event.report_descriptors) == 0:
```
[the remark this answers is not in the document]
Comment (jepebe) [reply toadr3/report.py:192]: Yes, but I have shot myself in the foot with that one before. 😬

File: toadr3/reports.py (hunk 4 above)
Comment (jepebe) [reply toadr3/reports.py:46]: The magic happens earlier:

```
async with aiohttp.ClientSession(json_serialize=toadr3.toadr_json_serialize) as session:
```

File: toadr3/report.py (hunk 3 above)
Comment (jepebe) [reply toadr3/report.py:206]: Agree!

File: toadr3/report.py (hunk 2 above)
Comment (jepebe) [reply toadr3/report.py:196]: Maybe. This is part of the assumptions caveat for this PR. :)

File: toadr3/report.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (millitza) [reply toadr3/report.py:201]: Haha we all have.

File: toadr3/reports.py (hunk 4 above)
Comment (millitza) [reply toadr3/reports.py:46]: That makes sense!

**Review by millitza — APPROVED** [review]

**Pull request merged**